src/btree/col_srch.c

```diff
@@ -286,11 +286,37 @@ __wt_col_search(
      * because column-store files are dense, but in this case the caller searched past the end of
      * the table.
      */
-    cbt->ins_head = WT_COL_APPEND(page);
-    if ((cbt->ins = __col_insert_search(cbt->ins_head, cbt->ins_stack, cbt->next_stack, recno)) ==
-      NULL)
+    ins_head = WT_COL_APPEND(page);
+    ins = __col_insert_search(ins_head, cbt->ins_stack, cbt->next_stack, recno);
+    if (ins == NULL) {
+        /*
+         * Set this, otherwise the code in cursor_valid will assume there's no on-disk value
+         * underneath ins_head.
+         */
+        if (page->type != WT_PAGE_COL_FIX)
+            F_SET(cbt, WT_CBT_VAR_ONPAGE_MATCH);
+
+        /*
+         * There is nothing on the append list, so search the insert list. (The append list would
+         * have been closer to the search record).
+         */
+        if (cbt->recno != WT_RECNO_OOB) {
+            ins_head = (page->type == WT_PAGE_COL_FIX) ? WT_COL_UPDATE_SINGLE(page) :
+                                                         WT_COL_UPDATE_SLOT(page, cbt->slot);
+            ins = WT_SKIP_LAST(ins_head);
+            if (ins != NULL && cbt->recno == WT_INSERT_RECNO(ins)) {
+                cbt->ins_head = ins_head;
+                cbt->ins = ins;
+            }
+        }
+
         cbt->compare = -1;
-    else {
+    } else {
+        if (page->type != WT_PAGE_COL_FIX)
```

Comment:
Why we need to set the flag here? In this code path, we return something from the append list so there should be no onpage value.

Reply:
I'm actually clearing the flag, so I think we actually agree?

Reply:
Oh, I see. But why we need to clear that if we don't set it before past end?

Reply:
We don't necessarily *need* to set it here, but it should be unset by the time we hit this scenario, so I unset it out of paranoia. An assert here would also be an appropriate tool if you'd prefer that?

Reply:
I would prefer an assert so we can catch anything unusual.

Reply:
Thanks, done in e8a254c.

Reply:
We should add the if check to the assert
WT_ASSERT(session, page->type == WT_PAGE_COL_FIX || !F_ISSET(cbt, WT_CBT_VAR_ONPAGE_MATCH));

Reply:
Is there a technical reason for that, or just style? I think the current approach is a little easier to read.

Reply:
I don't want the if check build into production build.

Reply:
I don't think it'll change anything, if `HAVE_DIAGNOSTIC` isn't defined then it just becomes

```
if (page->type != WT_PAGE_COL_FIX)
    (void)(session);
```

which will definitely get optimised out.

Reply:
yes, I agree. But we don't do this in the code.

Reply:
Fair enough. I disagree on the optimisation thing, but it seems to be house style to do it this way, so I've changed it in 8b92d03.